core/query/src/main/java/org/eclipse/rdf4j/query/QueryResult.java

```diff
@@ -18,11 +20,41 @@
  * @author Jeen Broekstra
  * @author Arjohn Kampman
  */
-public interface QueryResult<T> extends CloseableIteration<T, QueryEvaluationException>, Iterable<T> {
+public interface QueryResult<T>
+		extends AutoCloseable, CloseableIteration<T, QueryEvaluationException>, Iterable<T> {
 
 	@Override
 	default Iterator<T> iterator() {
 		return new CloseableIterationIterator<>(this);
 	}
 
+	/**
+	 * Returns {@code true} if the query result has more elements. (In other words, returns {@code true} if
+	 * {@link #next} would return an element rather than throwing a {@link NoSuchElementException}.)
+	 *
+	 * @return {@code true} if the iteration has more elements.
+	 * @throws QueryEvaluationException if an error occurs while executing the query.
+	 */
+	boolean hasNext() throws QueryEvaluationException;
```

Comment:
Since QueryEvaluationException is a runtime exception would it be allowed to drop the `throws` declaration? Or would japicmp complain?

Reply:
I guess we could, but I left it in to make it clear that this is an exception type that could happen. Though perhaps having that only in the javadoc is enough.

Reply:
Actually I think (haven't tested) that we still need this because it overrides a generic `hasNext() throws X` method in the superclass.

Reply:
That would make sense.